daemon/oci.c

```diff
@@ -0,0 +1,1053 @@
+			WARN("Failed to execute CREATE_RUNTIME hook!");
+	}
+	return 0;
+}
+
+int
+oci_do_hooks_create_container(const container_t *container)
+{
+	oci_container_t *oci_container = oci_get_oci_container_by_container(container);
+
+	for (list_t *l = oci_container->hook_create_container_list; l; l = l->next) {
+		oci_hook_t *hook_create_container = l->data;
+		if (-1 == oci_do_hook(hook_create_container, container))
+			WARN("Failed to execute CREATE_CONTAINER hook!");
+	}
+	return 0;
+}
+
+int
+oci_do_hooks_start_container(const container_t *container)
+{
+	oci_container_t *oci_container = oci_get_oci_container_by_container(container);
+
+	for (list_t *l = oci_container->hook_start_container_list; l; l = l->next) {
+		oci_hook_t *hook_start_container = l->data;
+		if (-1 == oci_do_hook(hook_start_container, container))
+			WARN("Failed to execute START_CONTAINER hook!");
+	}
+	return 0;
+}
+
+int
+oci_do_hooks_poststart(const container_t *container)
+{
+	oci_container_t *oci_container = oci_get_oci_container_by_container(container);
+
+	for (list_t *l = oci_container->hook_poststart_list; l; l = l->next) {
+		oci_hook_t *hook_poststart = l->data;
+		if (-1 == oci_do_hook(hook_poststart, container))
+			WARN("Failed to execute POSTSTART hook!");
+	}
+	return 0;
+}
+
+int
+oci_do_hooks_poststop(const container_t *container)
+{
+	oci_container_t *oci_container = oci_get_oci_container_by_container(container);
+
+	for (list_t *l = oci_container->hook_poststop_list; l; l = l->next) {
+		oci_hook_t *hook_poststop = l->data;
+		if (-1 == oci_do_hook(hook_poststop, container))
+			WARN("Failed to execute POSTSTOP hook!");
+	}
+	return 0;
+}
+
+oci_container_t *
+oci_get_oci_container_by_container(const container_t *container)
+{
+	for (list_t *l = oci_containers_list; l; l = l->next) {
+		oci_container_t *oci_container = l->data;
+		if (container == oci_container->container)
+			return oci_container;
+	}
+	return NULL;
+}
+
+static container_t *
+oci_control_get_container_by_id_string(const char *id_str)
+{
+	uuid_t *uuid = uuid_new(id_str);
+	if (!uuid) {
+		WARN("Could not get UUID");
+		return NULL;
+	}
+	container_t *container = cmld_container_get_by_uuid(uuid);
+	if (!container) {
+		WARN("Could not find container for UUID %s", uuid_string(uuid));
+		uuid_free(uuid);
+		return NULL;
+	}
+	uuid_free(uuid);
+	return container;
+}
+
+void
+oci_container_free(oci_container_t *oci_container)
+{
+	IF_NULL_RETURN(oci_container);
+
+	oci_containers_list = list_remove(oci_containers_list, oci_container);
+
+	// do not free oci_container->container, this is done by cmld module
+
+	mem_free0(oci_container->id);
+	mem_free0(oci_container->bundle);
+
+	for (list_t *l = oci_container->hook_prestart_list; l; l = l->next)
+		oci_hook_free(l->data);
+
+	for (list_t *l = oci_container->hook_create_runtime_list; l; l = l->next)
+		oci_hook_free(l->data);
+
+	for (list_t *l = oci_container->hook_create_container_list; l; l = l->next)
+		oci_hook_free(l->data);
+
```

Comment:
more hooks to be freed here?

	list_t *hook_start_container_list;
	list_t *hook_poststart_list;
	list_t *hook_poststop_list;

Reply:
added